### PR TITLE
Perform fragile layout for stored properties in classes

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <rdar://problem/38506059> and <https://bugs.swift.org/browse/SR-7206>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
```
Assertion failed: (V[I]->getType() == T->getTypeAtIndex(I) && "Initializer for composite element doesn't match!"), function ConstantAggregate, file /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/llvm/lib/IR/Constants.cpp, line 878.
21:04:22 0  swift                    0x000000010b46a9e8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
21:04:22 1  swift                    0x000000010b46b076 SignalHandler(int) + 422
21:04:22 2  libsystem_platform.dylib 0x00007fff4fd30f5a _sigtramp + 26
21:04:22 3  libsystem_malloc.dylib   0x00007fff4fc58384 tiny_malloc_from_free_list + 431
21:04:22 4  libsystem_c.dylib        0x00007fff4fb5b312 abort + 127
21:04:22 5  libsystem_c.dylib        0x00007fff4fb23368 basename_r + 0
21:04:22 6  swift                    0x000000010b292a7f llvm::ConstantAggregate::ConstantAggregate(llvm::CompositeType*, llvm::Value::ValueTy, llvm::ArrayRef<llvm::Constant*>) + 447
21:04:22 7  swift                    0x000000010b2a16f7 llvm::ConstantUniqueMap<llvm::ConstantStruct>::create(llvm::StructType*, llvm::ConstantAggrKeyType<llvm::ConstantStruct>, std::__1::pair<unsigned int, std::__1::pair<llvm::StructType*, llvm::ConstantAggrKeyType<llvm::ConstantStruct> > >&) + 71
21:04:22 8  swift                    0x000000010b293df2 llvm::ConstantUniqueMap<llvm::ConstantStruct>::getOrCreate(llvm::StructType*, llvm::ConstantAggrKeyType<llvm::ConstantStruct>) + 386
21:04:22 9  swift                    0x000000010b293bdb llvm::ConstantStruct::get(llvm::StructType*, llvm::ArrayRef<llvm::Constant*>) + 203
21:04:22 10 swift                    0x0000000107afaa60 swift::irgen::emitConstantObject(swift::irgen::IRGenModule&, swift::ObjectInst*, swift::irgen::StructLayout*) + 560
21:04:22 11 swift                    0x0000000107bedaf5 swift::irgen::IRGenModule::emitSILStaticInitializers() + 453
21:04:22 12 swift                    0x0000000107b01649 swift::irgen::IRGenerator::emitGlobalTopLevel() + 601
21:04:22 13 swift                    0x0000000107bc77de performIRGeneration(swift::IRGenOptions&, swift::ModuleDecl*, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, swift::SourceFile*, llvm::GlobalVariable**, unsigned int) + 1198
21:04:22 14 swift                    0x0000000107bc613d swift::performIRGeneration(swift::IRGenOptions&, swift::ModuleDecl*, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule> >, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::LLVMContext&, llvm::ArrayRef<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, llvm::GlobalVariable**) + 829
21:04:22 15 swift                    0x0000000107a64e3f performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 15535
21:04:22 16 swift                    0x0000000107a6017e swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3310
21:04:22 17 swift                    0x0000000107a1e5b3 main + 2051
21:04:22 18 libdyld.dylib            0x00007fff4faaf115 start + 1
```

Sigh... this hack might not work out in practice.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/IRGen/GenClass.cpp [hunk 1]
```diff
@@ -245,6 +249,10 @@ namespace {
         }
 
         if (superclass->hasClangNode()) {
+          // Perform fragile layout if the class has @objc ancestry.
```
Comment (jrose-apple) [inline lib/IRGen/GenClass.cpp]: This comment's out of date since we ended up deciding not to conditionalize on that.

File: lib/IRGen/GenType.h [hunk 2]
```diff
@@ -220,6 +220,22 @@ class GenericContextScope {
...
 
+/// An RAII interface for forcing types to be lowered bypassing resilience.
+class CompletelyFragileScope {
+  TypeConverter &TC;
+public:
+  CompletelyFragileScope(TypeConverter &TC) : TC(TC) {
+    TC.pushCompletelyFragile();
+  }
+
+  CompletelyFragileScope(IRGenModule &IGM)
```
Comment (jrose-apple) [inline lib/IRGen/GenType.h]: Nitpick: can you make these constructors `explicit`?

**Review by jrose-apple — COMMENTED** [review]
I'm not an IRGen expert, but it looks good to me (well, as good as possible for this kind of workaround), and ready to go for the 4.2 branch. Glad the field access thing wasn't too bad.

File: lib/IRGen/GenClass.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/IRGen/GenClass.cpp]: The comment is still valid. Note that we only set CompletelyFragileLayout if the superclass has a Clang node.

File: lib/IRGen/GenType.h (hunk 2 above)
Comment (slavapestov) [reply lib/IRGen/GenType.h]: Oops, I usually remember to do that. I'll make a follow up patch.

File: lib/IRGen/GenClass.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/IRGen/GenClass.cpp]: Oh. That's not sufficient; you could be the grandchild of an imported class. I'm…sorry for not catching that sooner.

Comment (slavapestov) [reply lib/IRGen/GenClass.cpp]: That's fine. We walk all superclasses up to the root here.

**Pull request merged**